### ISSUE TITLE
[Bug] Single-device TTSampling returns wrong tokens: the iota indices_tensor it hands ttnn.topk is half the width of the logits

### ISSUE THREAD

**Issue by kinginu** [tenstorrent/tt-metal#52928]
**Component / Area:** models / sampling

**Issue Type:** Bug

## Observed

On a 1×1 mesh `TTSampling` returns tokens that are **not the model's argmax** whenever sampling takes the top-k path. Through `TTSampling.decode_forward` with `k=1`, 8 trials of 32 users, counting a token wrong only when its logit is strictly below the row maximum:

| padded vocab | half | top-k factory | wrong tokens |
|---|---|---|---|
| 1024 | 512 | single-core | 0/256 |
| 4096 | 2048 | single-core | 0/256 |
| 32768 | 16384 | **multi-core** | **126/256**, some outside the vocabulary |
| 65536 | 32768 | **multi-core** | **191/256** |

End to end on `mistralai/Mistral-7B-Instruct-v0.3`, `MESH_DEVICE=P150`, `simple_text_demo.py -k "performance-ci-token-matching"`, 500 teacher-forced tokens against the repo's own `reference_outputs/Mistral-7B-Instruct-v0.3.refpt`:

| sampling params | path taken | Top-1 | Top-5 |
|---|---|---|---|
| `temperature=0` — the config as committed | argmax fast path | 96.40% | 99.40% |
| `k=1, p=0.5, temp=1` | **top-k** | **83.20%** | **85.40%** |
| `k=1, p=0.5, temp=1`, with the fix below | top-k | **96.40%** | **99.40%** |

`p=0.5` is the whole difference between the first two rows: `_is_force_argmax_sampling` requires `p ∈ {0.0, 1.0}`, so `p=0.5` keeps the run on the top-k path while `k=1` keeps the result deterministic. The committed CI config uses `temperature=0`, which #48886 now routes to the argmax fast path on single-chip parts — which is why CI is green and this is not visible today.

This is also the root cause of #34763, open since December 2025, whose workaround disables on-device sampling for the whole Mistral-7B family.

### Root cause

Both samplers build a `[1, 1, 32, W]` buffer whose contents are `0..W-1` and hand it to `ttnn.topk` as `indices_tensor`.

`models/common/sampling/tt_sampling.py:342`:

```python
indices_tensor_torch = torch.zeros(1, 1, self.max_batch_size, padded_per_device, dtype=torch.int32)
for i in range(padded_per_device):
    indices_tensor_torch[:, :, :, i] = i
```

`padded_per_device` is `padded_vocab_size // _get_num_sampling_shards()`, and `_get_num_sampling_shards()` returns **2** on a 1×1 mesh (`tt_sampling.py:307`), so the buffer is `V/2` wide. `decode_forward` then splits it in half again and hands each `V/4` half to a `ttnn.topk` whose logits half is `V/2`. Every call passes an indices tensor half the width of its input.

Nothing catches that. `TopKDeviceOperation::validate_on_program_cache_miss` checks the indices tensor's *dtype* and never its shape against the input (`topk_device_operation.cpp:177`). Both readers then page the indices accessor with `page_id = i * Wt + j`, where `Wt` is the **input's** width in tiles (`reader_create_index_local_topk.cpp:65`, `reader_create_index_tensor.cpp`), so the second half of every row is read past the end of the buffer. The values are unaffected — they never come from that buffer — but the indices returned alongside them no longer belong to them. Feeding an 8192-wide iota to a 16384-wide input directly:

| | full-width iota | half-width iota |
|---|---|---|
| values | correct | identical |
| `gather(input, idx) == vals` | True | **False** |
| indices out of range | 0 | 274 |
| top-1 index wrong | 0/32 | 17/32 |

`topk_single_core_program_factory.cpp:212` hardcodes `GENERATE_INDICES = "1"` and ignores the argument entirely (#51921, #36329) — the comment at `:186` already records that "the reader's sole use sits behind `#if not GENERATE_INDICES` and that define is" never cleared. So the corruption only surfaces when a half reaches the multi-core factory: a power of two in `[8192, 65535)`.

### Which configurations reach it

On one device `padded_vocab_size = ceil(vocab/32)*32`, so the reachable vocabularies are 16384, 32768 and 65536. Of the models tt_transformers supports on a single card, only `mistralai/Mistral-7B-Instruct-v0.3` lands on one — 32768, halving to 16384. v0.1 and v0.2 are 32000, which halves to 16000 and falls back to the single-core factory.

Since #48886 set `allow_force_argmax` to `self.num_devices == 1`, single-chip **greedy** decode takes the argmax fast path and never reaches this. What is left exposed is **non-greedy sampling** on a single chip — any `top_k > 1`, any `top_p` other than 0 or 1, any `temperature` other than 1 — which is the ordinary vLLM path. #48886 narrowed the blast radius without touching the defect.

`Sampling1D` is not affected: its `_make_local_indices` already used the full width on the split path and records that as a fix over the TTTv1 port.

It survived because `TTSampling`'s 1×1 top-k path has no test at all. `models/common/tests/test_sampling.py` is TG-only, the gpt_oss and gemma4 unit tests are (4,8) and (1,8), and `test_sampling_1d.py` covers `Sampling1D`.

### Why this is #34763

`simple_text_demo.py:35` disables on-device sampling for the whole Mistral-7B family, pointing at #34763. #35153, which added that skip list, describes itself as a hack — *"Device sampling is not working with Mistral7B. For now skip sampling on device for Mistral7B. This needs proper fixing later."* Its author bisected the failure to `58ac27a9`, *"[TT-Transformers] Support on-device sampling (#31046)"* — the commit that introduced this buffer — and noted *"Seems like sampling on device was not tested for Mistral7b."* The last update on #34763 is 2026-01-07.

With the defect fixed, the skip list can come off: the same demo, skip list lifted, is 96.40% Top-1 on the committed greedy config (argmax fast path) and 96.40% on the top-k path, against 96.40% for today's host-sampling default. Decode is 24.78 ms/token today, 24.51 with device sampling on the greedy path.

## Expected

Drop the argument and the buffers rather than widening them. The buffer holds exactly what `generate_index_tile()` writes when the argument is absent, so the width the kernel generates for itself is the correct one by construction. Implemented and measured; a PR can follow.

**On today's main this costs a little at the op level and nothing at the model level.** `ttnn.topk`, `[1, 1, 32, W]`, `bfloat16`, 50 ops enqueued back to back, median of 30, the two branches interleaved in one process:

| W | read `indices_tensor` | generate | |
|---|---|---|---|
| 8192 | 96.93 | 107.84 | +11.3% |
| 16384 | 109.60 | 130.21 | +18.8% |
| 32768 | 128.37 | 186.95 | +45.6% |

`Sampling1D.decode_forward` on a 1×1 mesh, median of 30 samples of 10 calls, two interleaved runs: vocab 32768 goes 352.9 → 378.2 µs (+7.2%), vocab 1024 goes 269.2 → **264.9** µs (−1.6%, since the single-core path never read the buffer and the per-call `ttnn.split` and two `deallocate` go away).

At the model level that disappears into the step: Mistral-7B batch-1 decode on the top-k path is **24.71 ms/token before and 24.73 after**, +0.08%, which is noise. The sampler is roughly 1.4% of a decode step.

It also drops a resident `uint16` tensor per device — 1 MiB at a 16384-wide shard — re-read every decode step.

The op-level cost inverts with #52909, which builds the index tile with the NoC: with that change generation becomes the cheaper branch (105.83 µs against 109.81 at W=16384, −3.6%) and `Sampling1D.decode_forward` at vocab 32768 drops to 347.5 µs, −6.7% against today. That is a reason to want #52909, not a reason to hold this.

The per-shard offset that makes gathered indices global comes from `index_offsets` / `tt_indices_device_offsets`, a separate buffer, and is unaffected.

### Correctness

The generated index tile is bit-identical to the iota buffer it replaces. `ttnn.topk` with and without `indices_tensor`, values and indices compared bitwise and both compared against `torch.topk` on the bf16-rounded input, at W = 64, 128, 1024, 4096, 8192, 16384, 32768, 65536, 131072, 1536, 3000 and 64128 — both program factories, the UINT16→UINT32 index boundary at 65535, and non-power-of-2 widths. Identical on every check. The output index dtype is chosen from the input width inside the op, never from the passed tensor, so dropping the argument cannot change it.

`models/common/tests/modules/sampling/test_sampling_1d.py`: 92 passed, 48 skipped, identical before and after. `tests/ttnn/unit_tests/operations/reduce/test_topk.py`, `test_sampling.py` and `tests/ttnn/unit_tests/operations/data_movement/test_sort.py`: 345 passed, 8 skipped, 80 xfailed.

## Scope

| file | |
|---|---|
| `models/common/sampling/tt_sampling.py` | `_create_indices_tensors`, two `ttnn.topk` calls, `_select_topk_indices_dtype` becomes unused |
| `models/common/modules/sampling/sampling_1d.py` | `local_indices` config field and factory, two `ttnn.topk` calls |
| `models/common/tests/modules/sampling/test_sampling_1d.py` | one buffer assertion |
| `models/common/tests/test_sampling.py` | new 1×1 top-k-vs-argmax case at vocab 32768 and 32000, covering both factories; fails 21/32 users without the fix |
| `models/tt_transformers/demo/simple_text_demo.py` | drop `"Mistral-7B"` from `models_not_supported_for_device_sampling` |

`models/experimental/llama32_1b_quasar/` carries copies of both samplers with the same defect and is left alone; its vocab 128256 halves to 64128, so it is latent there.

The change leaves `ttnn.topk`'s parameter in place, but it is worth recording that afterwards **it has no production caller left in the repo**: the only remaining callers are two direct `ttnn.topk` calls in `test_sampling_1d.py` and the `pass_indices_tensor` parametrization in `tests/ttnn/unit_tests/operations/reduce/test_topk.py`, both of which test the op itself. The parameter is currently worse than unused — the multi-core factory reads it with no shape validation, and the single-core factory appends its `TensorAccessorArgs` and then ignores it. Deleting it removes roughly 60 lines across nine files and resolves #51921 by subtraction rather than by making the two factories agree. That is a public-API change and belongs in its own PR.

A shape check against the input in `validate_on_program_cache_miss` would have turned this into a loud failure instead of wrong tokens, and is worth adding whether or not the parameter survives.

## Environment

Blackhole P300, one chip visible, Ubuntu 22.04, main at `03b9d08219`, built Release. 1×1 and 1×2 meshes.

## Not established

* **Different part.** #34763's CI failures were N150 and N300; the reproduction above is Blackhole P150. Same code path, different silicon.
* **The N300 report is not explained by this defect.** On a 1×2 mesh the buffer width already matches the input width, and 1×2 measures 0 mismatches before and after the change. Either something else in the same feature is also broken, or the 2025-12-24 comment refers to a different test in the same job.
* 1×8 and 4×8 are unmeasured. 1×2 is the largest mesh available here, and multi-device is where the buffer is largest and the read most expensive.
* Whether the vLLM integration applies the same skip list is outside this repo. It is the path most exposed, since it is the one that uses real sampling parameters rather than greedy.

## Related

* #34763 — *Investigate mistral7b PCC and repeat batch errors*, the symptom.
* #35153 — the skip list added to contain it, self-described as a hack.
* #31046 / `58ac27a9` — introduced on-device sampling and this buffer.
* #48886 — routed single-chip greedy to the argmax fast path, which is why this is currently invisible in CI.
* #51921, #36329 — the single-core factory ignores `indices_tensor`.
* #25728, #36136 — added the buffer to take the read branch.
* #52909 — builds the index tile with the NoC; inverts the op-level cost above.
